Documented the local Symfony server (#11207)
symfony/symfony-docs: setup/symfony_server.rst

Symfony Local Web Server
========================

You can run Symfony applications with any web server (Apache, nginx, the

> internal PHP web server, etc.) However, Symfony provides its own web server to

Reviewer: either `However` should be a capital or there is missing a `.` here?

> make you more productive while developing your apps.

Reviewer: ```suggestion
make you more productive while developing your applications.
```

> This server is not intended for production use and it supports HTTP/2, TLS/SSL,

Reviewer 2: ```suggestion
Although this server is not intended for production use, it supports HTTP/2, TLS/SSL,
```

> Although this server is not intended for production use, it supports HTTP/2, TLS/SSL,

Reviewer 3: We should make this more prominent, that it is not for production use, with a `.. caution::` directive, WDYT @javiereguiluz ?

Author: I don't think it's needed ... and also the article title itself calls it "local web server".

automatic generation of security certificates, local domains, and many other
features that sooner or later you'll need when developing web projects.

> 

Reviewer 4: Would it make sense to mention that this web server is not tied to Symfony, but that one can use it on any PHP app and also on plain HTML/SPA applications?

Installation
------------

The Symfony server is distributed as a free installable binary without any
dependency and support for Linux, macOS and Windows. Go to `symfony.com/download`_
and follow the instructions for your operating system.

Getting Started
---------------

The Symfony web server is started once per project, so you may end up with
several instances (each of them listening to a different port). This is the
common workflow to serve a Symfony project:

.. code-block:: terminal

    $ cd my-project/

>     $ symfony server:start

Reviewer 4: You must add the `--no-tls` flag here to avoid an error message (TLS no configured yet at this point as you talk about it later on).

Author: I don't like that ... but I'll add it if it's really mandatory. What I'd like to have is this:

* `symfony server:start` allows browsing HTTP but not HTTPS (`--no-tls` is not necessary)
* If I try to browse `https://...` I see an informative error telling me that TLS is not enabled yet and that I should run `symfony server:ca:install` to enable it.

If you agree, is it possible to implement this? Thanks!

Reviewer 5: The problem is that you can't serve HTTPS traffic as you have no certificate generated yet.

Reviewer 5: But we can probably turn the error when not using `--no-tls` to a warning instead. WDYT?

Author: @tucksaun which one of these is more accurate?

1) Symfony server supports both HTTP and HTPPS. HTTP always works. HTTPS needs some previous work.
2) Symfony server only supports HTTPS. HTTP will never ever work.

I think 1) is true ... that's why this article is designed in a gradual way: 1) Install Server; 2) Use it (with HTTP); 3) Explain HTTPS, install it and use it for the rest of the article.

I don't want to show HTTPS as the first thing (... but I'll do if Symfony server doesn't work with HTTP).

Thanks.

Reviewer 5: It's more nuanced than this:
- Symfony server supports both HTTP and HTTPS.
- HTTP always works but is by default redirected to HTTPS once certificates are configured (one can use `--allow-http` with `symfony server:start` to prevent automatic redirection from HTTP to HTTPS).
- `server:start` *currently* requires the `server:ca:install` command to be run before except if ran with the `no-tls` flag.

We can probably chang the last fact so that `server:start` gets more easy to use for new setups and keep the automatic redirection to HTTPS by default.

Reviewer 5: @javiereguiluz the v4.3.3 version has been released, we now only issue a warning.

      [OK] Web server listening on http://127.0.0.1:....
      ...

    # Now, browse the given URL in your browser, or run this command:
    $ symfony open:local

Running the server this way makes it display the log messages in the console, so

> you can't run other commands. If you prefer, you can run the Symfony server in

Reviewer 2: ```suggestion
you won't be able to run other commands at the same time. If you prefer, you can run the Symfony server in
```

Author: I'm not questioning your fix because you are native in English and I'm not ... but this is confusing to me --> *"... at the same time."*

To me that means you can't run commands **simulataneously** ... but I don't want that. I may (or may not) run commands **later**, but if the terminal is blocked by Symfony's server, I can't.

So, it's still correct to add *"... at the same time."*  ? Thanks!

Reviewer 2: It should still be correct - 'at the same time' doesn't necessarily have to mean that you invoke them simutaneously, it just means that there is some concurrent activity.

Author: Fantastic! Thanks for fixing this ... and for teaching me about this expression!

the background:

.. code-block:: terminal

    $ cd my-project/

    # start the server in the background
    $ symfony server:start -d

    # continue working and running other commands...

    # show the latest log messages
    $ symfony server:log

Enabling TLS
------------

> Browsing the secure version of your apps locally is important to early detect

Reviewer 2: ```suggestion
Browsing the secure version of your apps locally is important to detect
```

> Browsing the secure version of your apps locally is important to early detect

Reviewer: ```suggestion
Browsing the secure version of your applications locally is important to early detect
```

> problems with mixed content and to run libraries that only run in HTTPS.

Reviewer 2: ```suggestion
problems with mixed content early, and to run libraries that only run in HTTPS.
```

Traditionally this has been painful and complicated to set up, but the Symfony
server automates everything. First, run this command:

.. code-block:: terminal

    $ symfony server:ca:install

This command creates a local certificate authority, registers it in your system
trust store, registers it in Firefox (this is required only for that browser)
and creates a default certificate for ``localhost`` and ``127.0.0.1``. In other
words, it does everything for you. You can now browse your local app using

> HTTPS instead of HTTP.

Reviewer 4: You need to restart existing servers.

Author: Could we do this automatically? Maybe like this:

```
$ symfony server:ca:install

  ...

  Servers must be restarted before using HTTPS. Do you want to restart them all now?
```

If `yes`:

```
  OK  All servers were restarted.
```

If `yes` but there was some problem:

```
  ERROR  Some servers couldn't be restarted automatically. Stop and start them manually.
```

If `no`:

```
  OK  Stop and start all the servers manually before using HTTPS.
```

> words, it does everything for you. You can now browse your local app using

Reviewer 3: ```suggestion
words, it does everything for you. You can now browse your local application using
```

Different PHP Settings Per Project
----------------------------------

Selecting a Different PHP Version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you have multiple PHP versions installed in your computer, you can tell

Reviewer: `in your computer` or `on your computer`?

Author: Should be "on" 😓

Symfony which one to use creating a file called ``.php-version`` at the project

> root dir:

Reviewer: ```suggestion
root directory:
```

.. code-block:: terminal

    $ cd my-project/

    # use a specific PHP version
    $ echo "7.2" > .php-version

    # use any PHP 7.x version available
    $ echo "7" > .php-version

> 

Reviewer 4: You might want to mention that the web server traverses the directory structure up to the root directory, so that's a great way to specify a default PHP version for ALL directories (`/.php-version`) of for a group of projects if they are under the same directory. I mention it as we have a quite a few questions about this on support.

> This other command is useful if you don't remember all the PHP versions

Reviewer 4: I would remove `other`

Reviewer 3: Or completely reword it like:

"To list all installed PHP versions you can execute...."

> installed in your computer:

Reviewer 2: ```suggestion
installed on your computer:
```

> installed in your computer:

Reviewer: `in` vs `on`

.. code-block:: terminal

>     $ symfony local:php:list

Reviewer 4: Probably interesting to mention that we list all supported SAPIs,  like CGI, FastCGI, ... which is important as having FastCGI installed is probably the best (as this is what you will have in production -- but not available on Windows). We favor FastCGI (php-fpm) then CGI (which acts as a FastCGI server as well), and finally, we fall back to plain CGI.

Overriding PHP Config Options Per Project
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can change the value of any PHP runtime config option per project creating a

Reviewer 2: ```suggestion
You can change the value of any PHP runtime config option per project by creating a
```

> file called ``php.ini`` at the project root dir. Add only the options you want

Reviewer: ```suggestion
file called ``php.ini`` at the project root directory. Add only the options you want
```

to override:

.. code-block:: terminal

    $ cd my-project/

    # this project only overrides the default PHP timezone

>     $ cat php.ini

Reviewer 3: this should be in a separate codeblock, maybe `.. code-block:: text`?

Author: This one is very tricky. If you show the contents of `php.ini` in a separate code block, the reading flow is broken too much. Even if it doesn't show highlighting, I think the current doc is the most obvious way to show that this is a file inside your project and these are its contents.

    [Date]
    date.timezone = Asia/Tokyo

Running Commands with Different PHP Versions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When running different PHP versions, it's useful to use the main ``symfony``

Reviewer 2: ```suggestion
When running different PHP versions it's useful to use the main ``symfony``
```

> command as a wrapper of the ``php`` command to always select the most

Reviewer 2: ```suggestion
command as a wrapper of the ``php`` command. This allows you to always select the most
```

> command as a wrapper of the ``php`` command. This allows you to always select the most

Reviewer 4: as a wrapper for?

appropriate PHP version according to the project which is running the commands:

.. code-block:: terminal

    # runs the command with the default PHP version
    $ php -r "..."

    # runs the command with the PHP version selected by the project
    # (or the default PHP version if the project didn't select one)

>     $ symfony php -r "..."

Reviewer 4: There are other benefits, like having the env vars automatically loaded (important when running non-Symfony commands).

If you are using this wrapper frequently, consider aliasing the ``php`` command
to it:

.. code-block:: terminal

    $ cd ~/.symfony/bin
    $ cp symfony php
    # now you can run "php ..." and the "symfony" command will be executed instead

Local Domain Names
------------------

> By default, projects are accessible at some random port of the ``12.7.0.0.1``

Reviewer: ```suggestion
By default, projects are accessible at some random port of the ``127.0.0.1``
```

> local IP. However, sometimes is preferable to associate a domain name to them:

Reviewer: ```suggestion
local IP. However, sometimes it is preferable to associate a domain name to them:
```

> local IP. However, sometimes is preferable to associate a domain name to them:

Reviewer 2: ```suggestion
local IP. However, sometimes it is preferable to associate a domain name to them:
```

* It's more convenient when you work continuously on the same project because
  port numbers can change but domains don't;

> * The behavior of some apps depend on their domains/subdomains;

Reviewer: ```suggestion
* The behavior of some applications depend on their domains/subdomains;
```

Reviewer 3: We use American English, so “behavior” is correct.

👍🏻 for applications

Reviewer: updated

> * To have stable endpoints, such as the local redirection URL of Oauth2.

Reviewer 4: for OAuth2?

Setting up the Local Proxy
~~~~~~~~~~~~~~~~~~~~~~~~~~

Local domains are possible thanks to a local proxy provided by the Symfony
server. First, start the proxy:

.. code-block:: terminal

    $ symfony proxy:start

If this is the first time you run the proxy, you must follow these additional steps:

* Open the **network configuration** of your operating system;
* Find the **proxy settings** and select the **"Automatic Proxy Configuration"**;

> * Set the following URL as its value: ``https://127.0.0.1:7080/proxy.pac``

Reviewer 4: I would add the image showing the window for the 3 operating systems (I can tweak the one I have in my slides for proper integration here).

Reviewer 5: good idea indeed

Author: Good idea! Sadly the Windows screenshot in your presentation is in French. We'll need it in English.

Defining the Local Domain
~~~~~~~~~~~~~~~~~~~~~~~~~

By default, Symfony proposes ``.wip`` (for *Work in Progress*) as the local
domains (but you can choose any other domain and TLD you like). Define a local
domain for a project as follows:

.. code-block:: terminal

    $ cd my-project/

>     $ symfony proxy:domain:attach my-domain.wip

Reviewer 4: You need to remove the `.wip` part.

If you have installed the local proxy as explained in the previous section, you

> can now browse ``https://my-domain.wip`` to access to your local project with

Reviewer: ```suggestion
can now browse ``https://my-domain.wip`` to access your local project with
```

the new custom domain.

.. tip::

    Browse the https://127.0.0.1:7080 URL to get the full list of local project
    directories, their custom domains, and port numbers.

When running console commands, add the ``HTTPS_PROXY`` env var to make custom
domains work:

.. code-block:: terminal

    $ HTTPS_PROXY=https://127.0.0.1:7080 curl https://my-domain.wip

Long-Running Commands
---------------------

Long-running commands, such as the ones related to compiling front-end web

> assets, block the terminal and you can't run other commands. The Symfony server

Reviewer 2: ```suggestion
assets, block the terminal and you can't run other commands at the same time. The Symfony server
```

provides a ``run`` command to wrap them as follows:

.. code-block:: terminal

    # compile Webpack assets using Symfony Encore ... but do that in the
    # background to not block the terminal
    $ symfony run -d yarn encore dev --watch

    # continue working and running other commands...

    # from time to time, check the command logs if you want
    $ symfony server:log

    # and you can also check if the command is still running
    $ symfony server:status
    Web server listening on ...
    Command "yarn ..." running with PID ...

>     # stop the command (and the whole server) when you are finished

Reviewer 4: `stop the web server (and all the associated commands)`?

    $ symfony server:stop

Bonus Features
--------------

> The Symfony server is much more than a local web server and it includes other

Reviewer 2: i'm not even sure this sentence is needed?

useful features.

Looking for Security Vulnerabilities
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Instead of installing the :doc:`Symfony Security Checker </security/security_checker>`
as a dependency of your projects, you can use this command from the Symfony server:

.. code-block:: terminal

    $ symfony security:check

This command uses the same vulnerability database as the Symfony Security
Checker but it also caches that information to keep checking security when

> it's not possible to access to that public database.

Reviewer 4: You can also add that it does not use the official API endpoint as everything is done locally (except for cloning the vulnerability DB). Which is probably the best in a CI scenario (adding a tip about this would be great).

> Creating Symfony Projects

Reviewer 3: I would propose using singular here:

Create a Symfony Project

Author: Not sure about this. I think it reads better in plural.

~~~~~~~~~~~~~~~~~~~~~~~~~

In addition to the `different ways to install Symfony`_, you can use this
command from the Symfony server:

.. code-block:: terminal

>     # creates a new project based on the Symfony Skeleton

Reviewer 3: ```suggestion
    # creates a new project based on symfony/skeleton
```

    $ symfony new my_project_name

>     # creates a new project based on the Symfony Website Skeleton

Reviewer 3: ```suggestion
    # creates a new project based on symfony/website-skeleton

    $ symfony new --full my_project_name

    # creates a new project based on the Symfony Demo application

>     $ symfony new --demo my_project_name

Reviewer 4: We can probably mention what that does behind the scenes as it's not just about running a `composer create-project` command.

Author: I don't understand this. The server doesn't make many extra things when creating new projects, right?

```bash
$ symfony new testing-project

* Creating a new Symfony project with Composer
  (running /usr/local/bin/composer create-project symfony/skeleton testing-project)

* Setting up the project under Git version control
  (running git init testing-project)

 [OK] Your project is now ready in testing-project/
```

SymfonyCloud Integration
------------------------

The local Symfony server provides full, but optional, integration with

> `SymfonyCloud`_, a service optimized to run your Symfony apps on the cloud.

Reviewer 3: ```suggestion
`SymfonyCloud`_, a service optimized to run your Symfony applications on the cloud.
```

It provides features such as creating environments, backups/snapshots, and
even access to a copy of the production data in your local machine to help

> you debug any issues.

Reviewer 3: ```suggestion
debug any issues.
```

Local Domain Names
------------------

By default, projects are accessible at some random port of the ``127.0.0.1``
local IP. However, sometimes it is preferable to associate a domain name to them:

* It's more convenient when you work continuously on the same project because
  port numbers can change but domains don't;
* The behavior of some applications depend on their domains/subdomains;
* To have stable endpoints, such as the local redirection URL for Oauth2.

Setting up the Local Proxy
~~~~~~~~~~~~~~~~~~~~~~~~~~

Local domains are possible thanks to a local proxy provided by the Symfony
server. First, start the proxy:

.. code-block:: terminal

    $ symfony proxy:start

If this is the first time you run the proxy, you must follow these additional steps:

* Open the **network configuration** of your operating system;
* Find the **proxy settings** and select the **"Automatic Proxy Configuration"**;
* Set the following URL as its value: ``https://127.0.0.1:7080/proxy.pac``

Defining the Local Domain
~~~~~~~~~~~~~~~~~~~~~~~~~

By default, Symfony proposes ``.wip`` (for *Work in Progress*) for the local

> domains (but you can choose any other TLD you like). Define a local domain for a

Reviewer 5: should we document how to change the TLD then?

file called ``php.ini`` at the project root directory. Add only the options you want
to override:

.. code-block:: terminal

    $ cd my-project/

    # this project only overrides the default PHP timezone
    $ cat php.ini
    [Date]
    date.timezone = Asia/Tokyo

Running Commands with Different PHP Versions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When running different PHP versions it's useful to use the main ``symfony``
command as a wrapper for the ``php`` command. This allows you to always select
the most appropriate PHP version according to the project which is running the
commands. It also loads the env vars automatically, which is important when
running non-Symfony commands:

.. code-block:: terminal

    # runs the command with the default PHP version
    $ php -r "..."

    # runs the command with the PHP version selected by the project
    # (or the default PHP version if the project didn't select one)
    $ symfony php -r "..."

If you are using this wrapper frequently, consider aliasing the ``php`` command
to it:

.. code-block:: terminal

    $ cd ~/.symfony/bin
    $ cp symfony php

>     # now you can run "php ..." and the "symfony" command will be executed instead

Reviewer 4: What about adding a note about the other wrappers that could be useful (`php-config`, `pear`, `pecl`, ...)?
